azure-pipelines.yml

```diff
@@ -17,69 +17,87 @@ jobs:
-        displayName: 'Download sbt 1.4.9'
-      - script: tar zxf /tmp/sbt.tgz -C /tmp/
-        displayName: 'Extract sbt'
-      - script: /tmp/sbt/bin/sbt ++2.12.8 clean
+      - script: sbt ++2.12.8 "project spark2" clean
         displayName: 'Running $sbt clean'
-      - script: /tmp/sbt/bin/sbt ++2.12.8 update
+      - script: sbt ++2.12.8 "project spark2" update
         displayName: 'Running $sbt update'
-      - script: /tmp/sbt/bin/sbt ++2.12.8 compile
+      - script: sbt ++2.12.8 "project spark2" compile
         displayName: 'Running $sbt compile'
-      - script: /tmp/sbt/bin/sbt ++2.12.8 test
+      - script: sbt ++2.12.8 "project spark2" test
         displayName: 'Running $sbt test'
       # If not a pull request, publish artifacts.
       - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
-        - script: /tmp/sbt/bin/sbt ++2.12.8 package
+        - script: sbt ++2.12.8 "project spark2" package
           displayName: 'Running $sbt package'
         - task: CopyFiles@2
           displayName: 'Copy hyperspace-core JAR'
           inputs:
             sourceFolder: '$(Build.SourcesDirectory)/target/'
             contents: '**/*.jar'
-            targetFolder: '$(Build.ArtifactStagingDirectory)/hyperspace-core/'
+            targetFolder: '$(Build.ArtifactStagingDirectory)/hyperspace-core_spark2.4/'
         - task: PublishBuildArtifacts@1
           displayName: 'Publish Hyperspace artifacts'
           inputs:
-            artifactName: 'hyperspace-core'
-            pathtoPublish: '$(Build.ArtifactStagingDirectory)/hyperspace-core/'
+            artifactName: 'hyperspace-core_spark2.4'
+            pathtoPublish: '$(Build.ArtifactStagingDirectory)/hyperspace-core_spark2.4/'
+
+  - job: Build_Spark3_2_12
+    displayName: 'Build sources and run unit tests for Spark 3.0 / Scala 2.12'
+    pool:
+      vmImage: 'ubuntu-18.04'
+    steps:
+      - script: sbt ++2.12.8 "project spark3" clean
+        displayName: 'Running $sbt clean'
+      - script: sbt ++2.12.8 "project spark3" update
+        displayName: 'Running $sbt update'
+      - script: sbt ++2.12.8 "project spark3" compile
+        displayName: 'Running $sbt compile'
+      - script: sbt ++2.12.8 "project spark3" test
+        displayName: 'Running $sbt test'
+      # If not a pull request, publish artifacts.
+      - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+        - script: sbt ++2.12.8 "project spark3" package
+          displayName: 'Running $sbt package'
+        - task: CopyFiles@2
+          displayName: 'Copy hyperspace-core JAR'
+          inputs:
+            sourceFolder: '$(Build.SourcesDirectory)/target/'
+            contents: '**/*.jar'
+            targetFolder: '$(Build.ArtifactStagingDirectory)/hyperspace-core_spark3.0/'
```

Comment:
Since we are heavily depending on the internal APIs, incompatibilities are very likely. I've already seen some incompatibilities between 3.0 and 3.1. It is our choice - we can either support only the latest minor version (e.g. when we move to 3.1 we drop the support for 3.0), or multiple minor versions.

The current scheme which includes minor version in the artifact name is more suitable if we are going to support multiple minor versions. There weren't huge differences between 2.4 and 3.0, so I think it is feasible to support multiple minor versions.

Reply:
Sounds good and we should support multiple minor versions. Let's get started on Spark 3.1 support once this PR is merged.

Reply:
Okay, then I'll make this PR more friendly to the future 3.1 support.